# METODO DE ROMBERG PARA APROXIMAR

# $$\int_a^b f(x)\,dx$$ 


## Tarea 2 
### Materia: Metodos Numéricos II
### Profesor: Ricardo Medel Esquivel 
### Alumno: Salazar Gonzalez Abel 

## Algoritmo 4.2 
En el codigo que hemos elaborado a continuación hemos programado el metodo de Romberg el cual hace uso de la regla compuesta del trapecio y la regla de Richardson para aproximar el calculo de integrales mediante la división del intervalo de integración en un número de intervalos arrojando todas las aproximaciones calculadas hasta arrojar la aproximación 
### $$R_{n,n}$$ 
la cual seria la extrapolacion con el nivel más alto de presición 

Para comprobar el buen funcionamiento del codigo se usara como ejemplo la siguiente integral resuelta anteriormente en el ejemplo de la página 160 de Burden

## $$\int_{0}^{\pi} \sin(x) \, dx$$
con n=6

Obteniendo asi los siguientes resultados 

|  |  |  |  |  |  |  |
| :--- | :--- | :--- | :--- | :--- | :--- | :--- |
| Fila 1 | 0.00000000 | | | | | |
| Fila 2 | 1.57079633 | 2.09439511 | | | | |
| Fila 3 | 1.89611890 | 2.00455976 | 1.99857073 | | | |
| Fila 4 | 1.97423160 | 2.00026917 | 1.99998313 | 2.00000555 | | |
| Fila 5 | 1.99357034 | 2.00001659 | 1.99999975 | 2.00000001 | 1.99999999 | |
| Fila 6 | 1.99839336 | 2.00000103 | 2.00000000 | 2.00000000 | 2.00000000 | 2.00000000 |

In [3]:
import math
import sympy as sp

x = sp.Symbol('x')
expresion_str = input("Ingrese la función f(x) a integrar: ")
expr = sp.sympify(expresion_str)
f = sp.lambdify(x, expr, modules=['math'])

a = float(sp.sympify(input("Valor del límite inferior a: ")))
b = float(sp.sympify(input("Valor del límite superior b: ")))
n = int(input("Número de niveles/filas n: "))

print("El valor de a es:", a)
print("El valor de b es:", b)
print("El valor de n es:", n)

# Paso 1: Obteneremos h inicial y el primer elemento R(1,1)
h = b - a
R1 = [0.5 * h * (f(a) + f(b))]

# Paso 2: El programa muestra a R[1.1]
print(f"\nFila 1: R[1,1] = {R1[0]}")

# Paso 3: Ciclo for para las filas i desde 2 hasta n
for i in range(2, n + 1):
    R2 = [0] * i  # Creamos la fila 2 para almacenar los calculos que el programa vaya haciendo
    
    # Paso 4: Cálculo del término R[2,1] usando la regla del trapecio
    sumatoria = 0
    limite_k = 2**(i - 2)
    for k in range(1, limite_k + 1): #ciclo for que ira por cada termino k de la sumatoria
    #y lo ira sustituyendo en las sumatorias de la funcion f  en la formla para obtener R[2,1]
        sumatoria += f(a + (k - 0.5) * h)
        
    R2[0] = 0.5 * (R1[0] + h * sumatoria)
    
    # Paso 5: Extrapolación de Richardson para calcular el resto de la fila
    for j in range(2, i + 1):
        R2[j-1] = R2[j-2] + (R2[j-2] - R1[j-2]) / (4**(j - 1) - 1)
    
    # Paso 6: Mostramos la fila calculada
    valores_fila = ", ".join([f"R[{i},{j}] = {R2[j-1]}" for j in range(1, i + 1)])
    print(f"Fila {i}: {valores_fila}")
    
    # Paso 7: Reducimos h a la mitad
    h = h / 2
    
    # Paso 8: Actualizamos la Fila 1 con los valores de la Fila 2 para la siguiente iteración
    R1 = R2

NEGRITA = "\033[1m"
RESET = "\033[0m"

# Resultado final (último valor calculado en la tabla)
resultado_final = R1[-1]
print(f"\nResultado de la integración: {NEGRITA}{resultado_final}{RESET}")


Ingrese la función f(x) a integrar:  sin(x)
Valor del límite inferior a:  0
Valor del límite superior b:  pi
Número de niveles/filas n:  6


El valor de a es: 0.0
El valor de b es: 3.141592653589793
El valor de n es: 6

Fila 1: R[1,1] = 1.9236706937217898e-16
Fila 2: R[2,1] = 1.5707963267948966, R[2,2] = 2.0943951023931953
Fila 3: R[3,1] = 1.8961188979370398, R[3,2] = 2.0045597549844207, R[3,3] = 1.9985707318238357
Fila 4: R[4,1] = 1.974231601945551, R[4,2] = 2.000269169948388, R[4,3] = 1.999983130945986, R[4,4] = 2.000005549979671
Fila 5: R[5,1] = 1.9935703437723395, R[5,2] = 2.0000165910479355, R[5,3] = 1.999999752454572, R[5,4] = 2.0000000162880416, R[5,5] = 1.9999999945872902
Fila 6: R[6,1] = 1.9983933609701445, R[6,2] = 2.0000010333694127, R[6,3] = 1.9999999961908446, R[6,4] = 2.0000000000596745, R[6,5] = 1.9999999999960338, R[6,6] = 2.000000000001321

Resultado de la integración: 2.000000000001321


Obteniendo de forma directa el resultado de la integral tenemos que 
#### $$\int_{0}^{\pi} \sin(x) \, dx =2 $$ 
Obteniendo el error entre la integración directa y por el metodo de simpson tenemos que 
#### $$E = |2 - 2.000000000001321 |\times 100 = 0.000000000132 $$
Por lo tanto podemos decir que obtenemos un resultado casi exacto al de la integración real 